### PR TITLE
Python installation

### PR DESCRIPTION
**Pull request by Danzelot**
Describe different ways of running python and installing packages.

### CODE REVIEW

**Comment by Danzelot** [comment]
@Chroxvi  Could you have look if it looks alright for you? I merged the wrapper section and I think it's okay for production now. Then we can merge your sections about using afterwards.

**Review by olouant — COMMENTED** [review]
I don't like this changes:

- Future plan is to use the the container wrapper not only for Python but also for R packages. The page should stay where it was
- With this structure you don't have the main navigation (on the left) for the cray-python and container wrapper pages

File: docs/software/installing/python.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    is horrible outdated for normal standards.
+
+## Background
+
+Python packages often consists of a lot of files. Especially bigger Conda
+environments often contain tens to hundreds of thousands of files.
+This puts a lot of strain on our LUSTRE file system which excels
+at serving serving bigger files to several nodes at the same time but struggles
+with many small files as the distribution of the metadata becomes a bottleneck.
+Therefore to ensure good file system performance, we enforce quite tight [quotas
```
Comment (olouant) [inline docs/software/installing/python.md:21]: Broken internal link

File: docs/software/installing/python/cray_python.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    to compile your own Python, C/C++ or Fortran code to use with HPE Cray
+    Python, you should ensure that you compile using `PrgEnv-gnu` to make sure
+    they are compatible.
+
+## Adding your own packages
+
+If the packages you require are not included in the HPE Cray Python
+distribution, further packages can be added using `pip`.  We recommend either
+installing them globally for your user account, as described in this section,
+or into virtual environments, see [next section](## Setting up virtual
```
Comment (olouant) [inline docs/software/installing/python/cray_python.md]: another broken internal link

Comment (Danzelot) [reply docs/software/installing/python/cray_python.md]: Ah, thanks for checking. Did you discover by hand or does mkdocs warn you about them? I got no warnings for those two.

Comment (olouant) [reply docs/software/installing/python/cray_python.md]: I test all the links, by hands :)

**Comment by Chroxvi** [comment]
> @Chroxvi Could you have look if it looks alright for you? I merged the wrapper section and I think it's okay for production now. Then we can merge your sections about using afterwards.

I actually like the idea of reducing the set of recommended approaches to installing/using Python on LUMI to just two:
1. Using Cray Python if you want an installation built against Cray's scientific libraries and do not need more packages than scipy, mpi4py, dask, and numpy.
2. Using a Singularity/Apptainer container for all other cases.

If we want to stick to just those two approaches, I think we need to restructure the documentation a bit. Specifically, I suggest we:

1. Update `installing/python.md` to just mention these two approaches (likely with a short warning against using the operating system Python and a plain conda/pip install) and then link to sub-pages detailing the two approaches further.
2. Update `installing/python/cray_python.md` to stress that we suggest that one only uses cray-python as-is without installing additional packages. However, if one is sure that one absolutely needs a cray-python based install, then one can use the approaches currently described in `installing/python/cray_python.md`.
3. Update `installing/python/container_self.md` to describe that we have two approaches to using a Python container: 1. Bring your own container, 2. Use the container wrapper (which I still believe that we should NOT recommend for production at this point). Furthermore, we can add all the technical details about why a container is needed over just a plain conda/pip install to `installing/python/container_self.md`.
4. Delete `installing/python/container_wrapper.md`.

@Danzelot, what do you think about this approach?

**Comment by Chroxvi** [comment]
Also, there seems to be something broken about the way `installing/python/cray_python.md` and `installing/python/container_self.md` are currently linked in the documentation. At least when using the preview from https://lumi-supercomputer-docs-preview.rahtiapp.fi/origin/python_installation/software/installing/python/, I get a timeout when following the links from `installing/python.md`. Is it a requirement to list such sub-pages in `mkdocs.yml`?

File: docs/software/installing/python.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    `python` without loading a module or container first will result in you using
+    the operating system default Python which is likely not what you intend as it
+    is horrible outdated for normal standards.
+
+## Background
+
+Python packages often consists of a lot of files. Especially bigger Conda
+environments often contain tens to hundreds of thousands of files.
+This puts a lot of strain on our LUSTRE file system which excels
+at serving serving bigger files to several nodes at the same time but struggles
```
Comment (rsarm) [inline docs/software/installing/python.md:19]:
```suggestion
at serving bigger files to several nodes at the same time but struggles
```

File: docs/software/installing/python.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+with many small files as the distribution of the metadata becomes a bottleneck.
+Therefore to ensure good file system performance, we enforce quite tight [quotas
+on the number of files](quota) which makes some common python
+installation methods impractical.
+
+## Overview of Python installation approaches
+
+Currently we recommend one of following approaches for installing Python
```
Comment (rsarm) [inline docs/software/installing/python.md:27]:
```suggestion
Currently we recommend one of the following approaches for installing Python
```

File: docs/software/installing/python/cray_python.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+The HPE Cray distribution provides Python 3 along with some of the most
+common packages used for scientific computation and data analysis. These
+include:
+
+  - numpy and scipy - built using GCC against HPE Cray LibSci
+  - mpi4py - built using GCC against HPE Cray MPICH
+  - dask
+
+The HPE Cray Python distribution can be loaded (either on the front-end
+or in a submission script) using:
```
Comment (rsarm) [inline docs/software/installing/python/cray_python.md:12]: Maybe writing 'The HPE Cray Python distribution can be loaded using' is enough? And also, de we need to load `LUMI/21.12` in order to load `cray-python`?

File: docs/software/installing/python/cray_python.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+installing them globally for your user account, as described in this section,
+or into virtual environments, see [next section](#setting-up-virtual-environments).
+
+Installing packages for your user account is in principle as simple as calling
+`pip`.  However, as the /home file systems are not available on the compute
+nodes, you will need to modify the default install location that `pip` uses to
+point to a location on the /work file systems (by default, `pip` installs into
+`$HOME/.local`). To do this, you set the `PYTHONUSERBASE` environment variable
+to point to a location on /work, for example:
+
```
Comment (rsarm) [inline docs/software/installing/python/cray_python.md:40]: Is this right? I think `$HOME` is available on the compute nodes. I think this section on changing where `pip install --user` installs the packages might not be needed.

File: docs/software/installing/python/cray_python.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+This uses the `--user` flag to ensure the packages are installed in
+the directory specified by `PYTHONUSERBASE`.
+
+## Setting up virtual environments
+
+Sometimes, you may need several different versions of the same package
+installed, for example, due to dependency issues. Virtual environments allow
+you to manage these conflicting requirements.  We recommend that you use `venv`
+to manage your Python environments. A summary of how to get a virtual
+environment set up is contained in the below, but for further information, see:
```
Comment (rsarm) [inline docs/software/installing/python/cray_python.md:77]:
```suggestion
environment set up is described below, but for further information, see:
```

File: docs/software/installing/python/cray_python.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+available within this environment. When running code that requires these
+packages you must activate the environment, by adding the above `source ...
+activate` line of code to any submission scripts.
+
+In case a newer version than the one provided by CRAY is needed you can also
+upgrade it with `pip install -U pandas`. But be aware that the new package
+version will not contain any optimizations for LUMI. So this is only
+recommended if the CRAY provided version is missing features or contains bugs
+fixed in newer versions.
+
```
Comment (rsarm) [inline docs/software/installing/python/cray_python.md:113]: I don't know if this is accurate. Not everything that comes preinstalled with `cray-python` has special cray stuff in it. Numpy in particular, I think it's only compiled against libsci and things that build on it, like pandas, will use that.

**Comment by Chroxvi** [comment]
The placement of this installing Python documentation in the LUMI Docs tree needs to be updated according to the new structure detail in https://wiki.eduuni.fi/x/cpRzEQ. We currently aim for a launch of this updated LUMI docs structure by end of August / beginning of September. This PR should be made ready for merge by then - but can't (easily) be merged sooner than that.

**Pull request closed without merging**